addr_manager.go

```diff
@@ -50,64 +80,94 @@ func (e *expiringAddr) ExpiredBy(t time.Time) bool {
 
 type addrSet map[string]expiringAddr
 
-// AddrManager manages addresses.
-// The zero-value is ready to be used.
 type AddrManager struct {
-	addrmu sync.Mutex // guards addrs
-	addrs  map[peer.ID]addrSet
-
+	addrmu   sync.Mutex
+	cache    *lru.Cache
+	dstore   ds.Datastore
 	addrSubs map[peer.ID][]*addrSub
 }
 
-// ensures the AddrManager is initialized.
-// So we can use the zero value.
-func (mgr *AddrManager) init() {
-	if mgr.addrs == nil {
-		mgr.addrs = make(map[peer.ID]addrSet)
+func newAddrManager(ctx context.Context, dstore ds.Batching) *AddrManager {
+	mgr := new(AddrManager)
+	mgr.addrSubs = make(map[peer.ID][]*addrSub)
+	cache, err := lru.New(lruCacheSize)
+	if err != nil {
+		panic(err) // only happens if negative value is passed to lru constructor
 	}
-	if mgr.addrSubs == nil {
-		mgr.addrSubs = make(map[peer.ID][]*addrSub)
+	mgr.cache = cache
+	mgr.dstore = autobatch.NewAutoBatching(dstore, batchBufferSize)
+	go mgr.cleanup(ctx)
+	return mgr
+}
+
+func (mgr *AddrManager) cleanup(ctx context.Context) {
+	tick := time.NewTicker(cleanupInterval)
+	for {
+		select {
+		case <-tick.C:
+			if err := mgr.deleteExpiredAddrs(); err != nil {
+				log.Error("failed to delete expired peer addresses from datastore:", err)
+			}
+		case <-ctx.Done():
+			tick.Stop()
+			return
+		}
 	}
 }
 
-func (mgr *AddrManager) Peers() []peer.ID {
+func (mgr *AddrManager) Peers() ([]peer.ID, error) {
 	mgr.addrmu.Lock()
 	defer mgr.addrmu.Unlock()
-	if mgr.addrs == nil {
-		return nil
-	}
 
-	pids := make([]peer.ID, 0, len(mgr.addrs))
-	for pid := range mgr.addrs {
-		pids = append(pids, pid)
+	res, err := mgr.dstore.Query(dsq.Query{
+		Prefix:   ds.NewKey(peerNamespace).String(),
+		KeysOnly: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	addrEnts, err := res.Rest()
```

Comment:
Take a look at the changes i *just* made to the providers storage: https://github.com/libp2p/go-libp2p-kad-dht/blob/master/providers/providers.go#L269

It turns out this pattern is actually pretty memory intensive...